File: src/main/java/com/vc/medium/L718.scala
package com.vc.medium

object L718 {
  def findLength(A: Array[Int], B: Array[Int]): Int = {
    val n = A.length
    val m = B.length
    val dp = Array.ofDim[Int](n + 1, m + 1)
    var res = 0
    (1 to n).foreach(i => {
      (1 to m).foreach(j => {
        if(A(i - 1) == B(j - 1)) {
          dp(i)(j) = dp(i - 1)(j - 1) + 1
          res = Math.max(res, dp(i)(j))
        }
      })
    })
    res
  }
}

File: src/main/java/com/vc/medium/L718.java
package com.vc.medium;

class L718 {
    public int findLength(int[] A, int[] B) {
        if(A == null||B == null) return 0;
        int m = A.length;
        int n = B.length;
        int max = 0;
        //dp[i][j] is the length of longest common subarray ending with nums[i] and nums[j]
        int[][] dp = new int[m + 1][n + 1];
        /**
             0 1 2 3 2 1
             0 0 0 0 0 0 0
             3 0 0 0 1 0 0
             2 0 0 1 0 2 0
             1 0 1 0 0 0 3
             4 0 0 0 0 0 0
             7 0 0 0 0 0 0
         */
        for(int i = 0; i <= m;i++){
            for(int j = 0;j <= n;j++){
                if(i == 0 || j == 0){
                    dp[i][j] = 0;
                }
                else{
                    if(A[i - 1] == B[j - 1]){
                        dp[i][j] = 1 + dp[i - 1][j - 1];
                        max = Math.max(max, dp[i][j]);
                    }
                }
            }
        }
        return max;
    }
}
